feat(datepicker): @Output for year and month selected in multiyear/year

julianobrasil opened the pull request:
`@Output`s to emit the selected year in multiyear view and the selected month in month view.

![image](https://thumbs.gfycat.com/ImpressiveGloriousCopperbutterfly-size_restricted.gif)

Closes  #9458

mmalerba commented on src/demo-app/datepicker/datepicker-demo.html:
```diff
@@ -133,3 +133,38 @@ <h2>Datepicker with value property binding</h2>
+
+<h2>Datepicker emulating year picker</h2>
+<p>
+  <input matInput [matDatepicker]="datePicker5" class="datepicker-input-invisible">
```
This hidden input stuff feels kind of hacky. What if we just use the `MomentDateAdapter` and provide a custom `MAT_DATE_FORMATS`

Our custom formats could just be a copy of [the defaults](https://github.com/angular/material2/blob/master/src/material-moment-adapter/adapter/moment-date-formats.ts) but with the parse and display formats for `dateInput` changed to `MM/YYYY` or `YYYY` respectively.

As for how to fit this into the demo, you could make some directives `demo-moment-month-year`, `demo-moment-year` that provide the appropriate things for their children

julianobrasil replied:
I'll work on it.

mmalerba commented on src/lib/datepicker/calendar.ts:
```diff
@@ -107,6 +107,18 @@ export class MatCalendar<D> implements AfterContentInit, OnDestroy, OnChanges {
 
+  /**
+   * Emits the year chosen in multiyear view.
+   * This doesn't implies in change on the selected date.
```
This doesn't imply a change in the selected date.

mmalerba commented on src/lib/datepicker/calendar.ts:
```diff
@@ -107,6 +107,18 @@ export class MatCalendar<D> implements AfterContentInit, OnDestroy, OnChanges {
+
+  /**
+   * Emits the month chosen in year view.
+   * This doesn't implies in change on the selected date.
```
This doesn't imply a change in the selected date.

mmalerba commented on src/lib/datepicker/datepicker.ts:
```diff
@@ -164,6 +164,18 @@ export class MatDatepicker<D> implements OnDestroy {
 
+  /**
+   * Emits selected year in multiyear view.
+   * This doesn't implies in change on the selected date.
```
same change to wording here

mmalerba commented on src/lib/datepicker/year-view.ts:
```diff
@@ -65,6 +65,9 @@ export class MatYearView<D> implements AfterContentInit {
   /** Emits when a new month is selected. */
   @Output() readonly selectedChange = new EventEmitter<D>();
 
+  /** Emits the selected month. This doesn't implies in change on the selected date */
```
same change to wording

mmalerba commented on src/lib/datepicker/multi-year-view.ts:
```diff
@@ -67,9 +67,12 @@ export class MatMultiYearView<D> implements AfterContentInit {
+  /** Emits a date when a new year is selected. */
   @Output() readonly selectedChange = new EventEmitter<D>();
 
+  /** Emits the selected year. This doesn't implies in change on the selected date */
```
same change to wording

julianobrasil commented:
Well, it's done. There are two things I'm not sure whether to touch:

1. A way to disable the period button that switch views: if the developer want to use this feature to emulate a year/month picker, maybe he/she also wish to avoid the end user changing views by clicking on that button.

2. The docs examples: in one comment above I've writen that maybe the the examples should be in a different PR, but I'm having second thoughts about it (because this PR would block the other). Do we have time for doing this part yet?

mmalerba commented:
1. I think we can save that for another PR (https://github.com/angular/material2/pull/9639), if we do it at all. There's another PR ongoing to allow specifying a custom header for the calendar which would allow people to address this issue.

2. I'll leave it up to you whether to do the docs in this PR or a separate one. If you want I can approve this one after all of the comments are addressed and you can start working on the docs. If you finish the docs before this gets merged we can just put it all in this PR.

julianobrasil commented on src/demo-app/datepicker/datepicker-demo.ts:
```diff
@@ -31,11 +36,88 @@ export class DatepickerDemo {
+  }
+
+  chosenYearFromYearMonthHandler(year: number) {
+    try {      
```
Yeah, that's something I put in the code and for no reason I left them there. I've just taken it away.

julianobrasil commented:
You can go on and aprove this one.

julianobrasil commented:
I have just rebased it.

mmalerba commented on src/demo-app/datepicker/datepicker-demo.ts:
```diff
@@ -31,11 +36,78 @@ export class DatepickerDemo {
   onDateChange = (e: MatDatepickerInputEvent<Date>) => this.lastDateChange = e.value;
 
-  dateCtrl = new FormControl();
+  chosenYearHandler(year: number, datepicker: MatDatepicker<Date>) {
```
nit: these should probably actually be `MatDatepicker<Moment>` (or `MatDatepicker<any>` since your code doesn't care)

julianobrasil replied:
I'm gonna use Moment for that just to expose the fact we're using Moment.

julianobrasil commented:
Just added docs example and text. I added a subsection in `Setting the calendar starting view`. Is it appropriate?

![image](https://user-images.githubusercontent.com/25405143/35673517-f48274f0-0728-11e8-9516-4e7b1ed6f172.png)

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,23 @@ year containing the `startAt` date.
 
+#### Watching the views for changes on selected years and months
+
+When a year or a month is selected in **multi-year and year views** respecively, the `yearSelected`
```
...in the `multi-year` and `year` views...

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,23 @@ year containing the `startAt` date.
+#### Watching the views for changes on selected years and months
+
+When a year or a month is selected in **multi-year and year views** respecively, the `yearSelected`
+and `monthSelected` outputs emit the chosen year (full year) and month (0 - 11).
```
now that I think about this some more, wouldn't it make sense to emit normalized dates
(e.g. `new Date(year, 0, 1)` and `new Date(year, month, 1)`)? especially for the month case, the month number alone doesn't give you all the info you need

julianobrasil replied:
ok, I'm gonna update this and push it tonight.

mmalerba commented on src/material-examples/datepicker-views-selection/datepicker-views-selection-example.html:
```diff
@@ -0,0 +1,7 @@
+<mat-form-field>
+  <input matInput [matDatepicker]="dp" placeholder="Choose a year and a month" [formControl]="date">
+  <mat-datepicker-toggle matSuffix [for]="dp" (yearSelected)="chosenYearHandler($event)"
```
These listeners should go on the datepicker, not the toggle

julianobrasil replied:
ops!

mmalerba commented:
I'm pretty happy with how this turned out. It's actually not that difficult to implement month and year selector with these outputs :)

julianobrasil commented:
I've changed the first paragraph of `datepicker.md`. But I'm not sure if it's clear so please  have a closer look at it.

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,29 @@ year containing the `startAt` date.
+#### Watching the views for changes on selected years and months
+
+When a year or a month is selected in `multi-year` and `year` views respecively, the `yearSelected`
+and `monthSelected` outputs emit the normalized chosen year and normalized month. By "normalized"
```
...outputs emit a normalized date representing the chosen year or month. By "normalized" we mean that the dates representing a year will have their month set to January and their day set to the 1st. Dates representing months will have their day set to the 1st of the month.  For example, if...

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,29 @@ year containing the `startAt` date.
+we mean that it will emit an object, not just a number. For example, if `<mat-datepicker>` is
+configured to work with javascript native Date objects, the `yearSelected` will emit
+`new Date(2017,0,1)` if the user selects 2017 in `multi-year` view. Similarly, `monthSelected`
+will emit `new Date(2017,1,0)` if the user selects **Jan** in `year` view and the current date
```
**Jan** --> January

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,29 @@ year containing the `startAt` date.
+configured to work with javascript native Date objects, the `yearSelected` will emit
+`new Date(2017,0,1)` if the user selects 2017 in `multi-year` view. Similarly, `monthSelected`
+will emit `new Date(2017,1,0)` if the user selects **Jan** in `year` view and the current date
+value of the connected `<input>` was something like `new Date(2017,MM,dd)` (the month and year
```
month and day

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,29 @@ year containing the `startAt` date.
+associated `<input>`.
+
+The following example uses `yearSelected` and `monthSelected` outputs to emulate a month and year
+picker (if you're not familiar with the usage of `MomentDateAdapter` and `formats customization`
```
`formats customization` ---> `MAT_DATE_FORMATS`

mmalerba commented on src/lib/datepicker/datepicker.md:
```diff
@@ -51,6 +51,29 @@ year containing the `startAt` date.
+
+The following example uses `yearSelected` and `monthSelected` outputs to emulate a month and year
+picker (if you're not familiar with the usage of `MomentDateAdapter` and `formats customization`
+you can read more about them below in this document to fully understand the example).
```
link the words "read more about them" to the correct header in this document.

mmalerba commented on tslint.json:
```diff
@@ -81,7 +81,7 @@
     "import-blacklist": [true, "rxjs", "rxjs/operators"],
     // Avoids inconsistent linebreak styles in source files. Forces developers to use LF linebreaks.
-    "linebreak-style": [true, "LF"],
+    "linebreak-style": [false, "LF"],
```
were the changes to this file made intentionally? I don't think we want to change these settings

julianobrasil replied:
Absoluto noy. I'm uses to switch those to false just to run tslint locally, and this time I forgot to switch them back after tests.

julianobrasil commented:
Just fixed  a small nit...

mmalerba commented:
@julianobrasil This is breaking google's build when I try to presubmit it because we don't sync the moment adapter into google3, but we try to verify that the demo app builds. Could you revert the changes to the demo app? (the examples are what's important to users anyways, demo app is just for the sake of our own manual testing)

julianobrasil commented:
I removed the demo changes, but there's still this error in the build process:

```
6 Feb 02:07:41 - error querying from https://saucelabs.com/rest/v1/angular-ci/tunnels/fa545dcef4e04d65927544a2ab05f0e3, error was: {"message": "Service is temporarily unavailable"}
```
